File: strategy/marvin/test/unit/cache.lua
local Cache = require "../base/cache"


local function foo(a, b, c)
	return a*(b+c)
end
foo = Cache.forFrame(foo)

local function bar()
	return 4
end
bar = Cache.forFrame(bar)

local function echo(...)
	return {...}
end
echo = Cache.forFrame(echo)


local function multiReturn()
	return 1, 2, 3
end
multiReturn = Cache.forFrame(multiReturn)

local function heavy()
	local a = 0
	for i = 0, 1000000 do
		a = a + i
	end
end
heavy = Cache.forFrame(heavy)

local side = 0
local function sideEffect()
	side = side + 1
end
sideEffect = Cache.forFrame(sideEffect)

context("base.cache", function ()
	test("different arguments", function ()
		local a = foo(1,2,3)
		local b = foo(2,3,4)
		assert_not_equal(a, b, "different arguments should result in different results")
	end)

	test("nil parameters", function ()
		-- unused and nil parameters should not pose problems (multiple calls are ok)
		local a = bar()
		local b = bar("bla")
		local c = bar(nil, 7)
		-- equal to a
		local d = bar(nil, nil, nil)
		assert_equal(a, 4)
		assert_equal(b, 4)
		assert_equal(c, 4)
		assert_equal(d, 4)
	end)

	test("parameters", function ()
		local a = echo()
		local b = echo("bla")
		local c = echo(nil, 7)
		local d = echo(nil, nil, nil, 5)
		assert_deep_equal(a, {})
		assert_deep_equal(b, { "bla" })
		assert_deep_equal(c, { nil, 7 })
		assert_deep_equal(d, { nil, nil, nil, 5 })
	end)

	test("side effects", function ()
		sideEffect()
		local before = side
		sideEffect()
		local after = side
		assert_equal(before, after, "when called with the same arguments, the function should only be called once")
	end)

	test("reset frame cache", function ()
		sideEffect()
		local before = side
		Cache.resetFrame()
		sideEffect()
		local after = side
		assert_equal(before + 1, after)
	end)

	test("multiple return", function ()
		local r1, r2, r3 = multiReturn()
		-- multiple return values should be possible
		assert_not_nil(r1)
		assert_not_nil(r2)
		assert_not_nil(r3)
	end)

	test("heavy", function ()
		-- some number-crunching for time-measuring
		for i = 1, 100000 do
			heavy()
		end
		assert_true(true)
	end)
end)

File: strategy/typescript/glados/test/unit/base/cache.ts
/**************************************************************************
*   Copyright 2026 Robotics Erlangen e.V., Tobias Heineken                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
**************************************************************************/

import * as Cache from "base/cache";

import { UnitTest } from "glados/test/unit/unittest";

export class BaseCache extends UnitTest {
	public constructor() {
		super();
		this._addTest("different arguments", this._testDifferentArguments);
		this._addTest("undefined parameters", this._testUndefinedParameters);
		this._addTest("parameters", this._testParameters);
		this._addTest("side effects", this._testSideEffects);
		this._addTest("heavy", this._testHeavy);
		this._addTest("forever", this._testForever);
		this._addTest("function name", this._testFunctionName);
	}

	private _testFunctionName() {
		function bar() {
			return 4;
		}
		function foo() {
			return 5;
		}
		let cachedBar = Cache.forFrame(bar);
		let cachedFoo = Cache.forFrame(foo);

		this._assert_eq(bar.name, cachedBar.name);
		this._assert_eq(foo.name, cachedFoo.name);
	}

	private _testDifferentArguments() {
		function foo(a: number, b: number, c: number) {
			return a * (b + c);
		}
		let cached = Cache.forFrame(foo);

		let a = cached(1, 2, 3);
		let b = cached(2, 3, 4);
		this._assert_ne(a, b);
	}

	private _testUndefinedParameters() {
		function bar() {
			return 4;
		}
		let cached: Function = Cache.forFrame(bar);

		// unused and undefined parameters should not pose problems (multiple calls are ok)
		let a = cached();
		let b = cached("bla");
		let c = cached(undefined, 7);
		// equal to a
		let d = cached(undefined, undefined, undefined);
		this._assert_eq(a, 4);
		this._assert_eq(b, 4);
		this._assert_eq(c, 4);
		this._assert_eq(d, 4);
	}

	private _testParameters() {
		function echo(...args: any[]) {
			return args;
		}
		let cached: any = Cache.forFrame(echo);

		let a = cached();
		let b = cached("bla");
		let c = cached(undefined, 7);
		let d = cached(undefined, undefined, undefined, 5);
		this._assert_deep_eq(a, []);
		this._assert_deep_eq(b, ["bla"]);
		this._assert_deep_eq(c, [undefined, 7]);
		this._assert_deep_eq(d, [undefined, undefined, undefined, 5]);
	}

	private _testSideEffects() {
		let side = 0;
		function sideEffect() {
			side = side + 1;
		}
		let cached = Cache.forFrame(sideEffect);

		cached();
		let before = side;
		cached();
		let after = side;
		this._assert_eq(before, after);

		Cache.resetFrame();
		cached();
		let afterReset = side;
		this._assert_eq(after + 1, afterReset);
	}

	private _testHeavy() {
		function heavy() {
			let a = 0;
			for (let i = 0; i < 1000000; i++) {
				a = a + i;
			}
		}
		let cached = Cache.forFrame(heavy);

		// some number-crunching for time-measuring
		for (let i = 0; i < 100000; i++) {
			cached();
		}
		this._assert_true(true);
	}

	private _testForever() {
		let side = 0;
		function sideEffect() {
			side = side + 1;
		}
		let cached = Cache.forever(sideEffect);

		let before = side;
		cached();
		let mid = side;
		cached();
		let after = side;
		this._assert_eq(before + 1, mid);
		this._assert_eq(mid, after);

		Cache.resetFrame();
		cached();
		let afterReset = side;
		this._assert_eq(after, afterReset);
	}
}
export let testClass = BaseCache;
